### PR TITLE
Search, part 3: add click-through analytics

### LINKED ISSUE #1877: Add click-through analytics to Algolia
## Background

We use Algolia for search, but there are a range of analytics about how search is performing that we can't get unless we also tell Algolia how people interacted with the search, in particular what they clicked on.

Image

This information could be important for the discovery phase of the Nav & Search Improvements project starting in March.

## Scope

Integrate an Algolia insights client and register the `clickedObjectIDsAfterSearch` event when people choose a search result.

## Technical notes

- API refererence: `clickedObjectIDsAfterSearch`: https://www.algolia.com/doc/api-reference/api-methods/clicked-object-ids-after-search/
- `userToken` best practices: https://www.algolia.com/doc/guides/sending-events/implementing/best-practices/usertoken/
  - We will need to ask Algolia to generate a per-user tracking cookie for search analytics, which don't happen by default in the recent client

### PR DESCRIPTION
**Pull request by marcelgerber**
Fixes #1877 | Live on _tufte_, together with #1915.

By communicating back click-through analytics to Algolia, they can provide us with insights like this one here:
image

As you can see, this is working fine already :)

One pretty big caveat right now is that for charts, we track click-throughs for all chart links, but don't track clicking on the interactive chart result, or interacting with it.

### CODE REVIEW

File: site/CookiePreferencesManager.tsx [hunk 1]
```diff
@@ -168,8 +168,9 @@ export const parseDate = (date?: string): number | undefined => {
 
 export const getPreferenceValue = (
     type: PreferenceType,
-    preferences: Preference[]
+    preferences?: Preference[]
 ) => {
+    preferences ??= getInitialState().preferences
```
Comment (ikesau) [inline site/CookiePreferencesManager.tsx:173]: Just checking, the only difference between this and a default parameter is that it coalesces `null` too?

**Review by ikesau — APPROVED** [review]
Nice and clean! 

Tested on tufte (with analytics on and off) and it appears to work correctly. (though I can't see your test environment in Algolia to examine the events)

I noticed when you click the chart preview title, nothing is tracked. It's an annoying amount of prop drilling to implement, but maybe we should check with Bobbie if that's something we'd like to do too.

EDIT: just noticed you specifically mentioned this in the OP 🤦 😅

File: site/CookiePreferencesManager.tsx (hunk 1 above)
Comment (marcelgerber) [reply site/CookiePreferencesManager.tsx]: Yeah, there was no real reason for that really :)
It's just that at the time I didn't think about default parameters, but they're of course nicer!

**Comment by marcelgerber** [comment]
> I noticed when you click the chart preview title, nothing is tracked. It's an annoying amount of prop drilling to implement, but maybe we should check with Bobbie if that's something we'd like to do too.

Yes, that's indeed the case.
I realize now that we could also add a catch-all `click` handler to the chart chrome. It gets tricky though: A click on the chart title is of course a pretty good indicator that the chart is interesting, but any other click inside the chart chrome (e.g. adding countries, changing tabs) is also a valuable interaction. And in many cases, the user will just look at the chart or maybe hover over it.
It becomes hard to track that well very quickly.

Given all of this, I vote not doing any tracking on the interactive chart, also given how we will overhaul search in a few months time.
This is a major drawback, of course, so we should openly document it.

**Comment by marcelgerber** [comment]
Also, another caveat that's of course always there with any tracking features is adblockers:
I use uBlock, and while it allows using Algolia just fine, it blocks any tracking requests we do.
So the CTRs etc. we get back will be undercounts.

**Pull request merged**